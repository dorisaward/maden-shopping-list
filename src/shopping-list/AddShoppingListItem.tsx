import React, {useCallback, useState} from 'react';
import {GroceryItem} from '../groceries/GroceryItem';
import {View, TextInput, TouchableOpacity, Text, Alert} from 'react-native';
import {styles} from '../styles';
import uuid from 'react-native-uuid';

type Props = {
  addGroceryItem: (groceryItem: GroceryItem) => void;
};

export const AddShoppingListItem = ({
  addGroceryItem,
}: Props): React.JSX.Element => {
  const [newGroceryName, setNewGroceryName] = useState<string | undefined>(
    undefined,
  );
  const [newGroceryQuantity, setNewGroceryQuantity] = useState<number>(0);

  const setQuantity = useCallback(
    (text: string) => {
      try {
        const number = Number(text);
        setNewGroceryQuantity(number);
      } catch (e) {
        console.warn(`Could not cast string ${text} to number`, e);
      }
    },
    [setNewGroceryQuantity],
  );

  const saveGroceryItem = useCallback(() => {
    if (!newGroceryName) {
      Alert.alert('Please make sure your groceries have a name');
      return;
    }
    if (!newGroceryQuantity) {
      Alert.alert(`Please specify how much ${newGroceryName} you want.`);
      return;
    }
    addGroceryItem({
      name: newGroceryName,
      quantity: newGroceryQuantity,
      id: uuid.v4().toString(),
    });
    setNewGroceryQuantity(0);
    setNewGroceryName(undefined);
  }, [newGroceryName, newGroceryQuantity, addGroceryItem]);

  return (
    <View style={styles.shoppingListItem}>
      <TextInput
        placeholder="Name"
        value={newGroceryName}
        style={styles.shoppingListItemInput}
        onChangeText={setNewGroceryName}
      />
      <TextInput
        placeholder="Quantity"
        value={newGroceryQuantity.toString()}
        keyboardType="number-pad"
        style={styles.shoppingListItemInput}
        onChangeText={setQuantity}
      />
      <TouchableOpacity style={styles.saveText} onPress={saveGroceryItem}>
        <Text>Save</Text>
      </TouchableOpacity>
    </View>
  );
};
