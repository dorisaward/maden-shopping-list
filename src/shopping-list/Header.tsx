import {Text, View} from 'react-native';
import {styles} from '../styles';
import React from 'react';

export const Header = (): React.JSX.Element => (
  <View style={styles.background}>
    <Text style={styles.headerText}>Mayden Shopping List</Text>
  </View>
);
