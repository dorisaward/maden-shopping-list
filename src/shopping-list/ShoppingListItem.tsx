import React from 'react';
import {ListRenderItem, Text, View} from 'react-native';
import {GroceryItem} from '../groceries/GroceryItem';
import {styles} from '../styles';

export const ShoppingListItem: ListRenderItem<GroceryItem> = ({item}) => (
  <View style={styles.shoppingListItem}>
    <Text style={styles.shoppingListItemText}>{item.name}</Text>
    <Text style={styles.shoppingListItemText}>{item.quantity}</Text>
  </View>
);
