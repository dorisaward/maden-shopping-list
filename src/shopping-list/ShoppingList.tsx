import React, { useState } from 'react';
import {FlatList, FlatListProps} from 'react-native';
import {ShoppingListItem} from './ShoppingListItem';
import {GroceryItem} from '../groceries/GroceryItem';

type Props = {
  initialGroceryItems: GroceryItem[];
};

export const ShoppingList = ({
  initialGroceryItems,
}: Props): React.JSX.Element => {
const [groceryItems, setGroceryItems] = useState<GroceryItem[]>(initialGroceryItems)

  return (
    <FlatList
      data={groceryItems}
      renderItem={ShoppingListItem}
      keyExtractor={keyExtractor}
    />
  );
};

const keyExtractor: FlatListProps<GroceryItem>['keyExtractor'] = item =>
  item.id;
