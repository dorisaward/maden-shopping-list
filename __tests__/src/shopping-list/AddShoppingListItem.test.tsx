import 'react-native';
import React from 'react';
import {it, expect} from '@jest/globals';
import {AddShoppingListItem} from '../../../src/shopping-list/AddShoppingListItem';
import {render, fireEvent} from '@testing-library/react-native';
import {Alert} from 'react-native';

it('should render add shopping list item', () => {
  // Given
  const addGroceryItem = jest.fn();
  const renderable = <AddShoppingListItem addGroceryItem={addGroceryItem} />;

  // When
  const {toJSON} = render(renderable);

  // Then
  expect(toJSON()).toMatchSnapshot();
});

it('should call add grocery item', () => {
  // Given
  const addGroceryItem = jest.fn();
  const renderable = <AddShoppingListItem addGroceryItem={addGroceryItem} />;
  const {getByText, getByPlaceholderText} = render(renderable);

  const groceryName = 'test name';
  fireEvent(getByPlaceholderText('Name'), 'onChangeText', groceryName);

  const groceryQuantity = 3;
  fireEvent(getByPlaceholderText('Quantity'), 'onChangeText', groceryQuantity.toString());

  // When
  fireEvent.press(getByText('Save'));

  // Then
  expect(addGroceryItem).toBeCalledWith({
    id: expect.any(String),
    name: groceryName,
    quantity: groceryQuantity,
  });
});

it('should not call add grocery item, given no name', () => {
  // Given
  jest.spyOn(Alert, 'alert');
  const addGroceryItem = jest.fn();
  const renderable = <AddShoppingListItem addGroceryItem={addGroceryItem} />;
  const {getByText} = render(renderable);

  // When
  fireEvent.press(getByText('Save'));

  // Then
  expect(Alert.alert).toBeCalledWith(
    'Please make sure your groceries have a name',
  );
});

it('should not call add grocery item, given no quantity', () => {
  // Given
  jest.spyOn(Alert, 'alert');
  const addGroceryItem = jest.fn();
  const renderable = <AddShoppingListItem addGroceryItem={addGroceryItem} />;
  const {getByText, getByPlaceholderText} = render(renderable);

  const groceryName = 'test name';
  fireEvent(getByPlaceholderText('Name'), 'onChangeText', groceryName);

  // When
  fireEvent.press(getByText('Save'));

  // Then
  expect(Alert.alert).toBeCalledWith(
    `Please specify how much ${groceryName} you want.`,
  );
});
