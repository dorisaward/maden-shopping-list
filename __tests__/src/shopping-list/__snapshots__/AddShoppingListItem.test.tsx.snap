// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add shopping list item 1`] = `
<View
  style={
    {
      "borderColor": "#6495ed",
      "borderRadius": 30,
      "borderWidth": 3,
      "flexDirection": "row",
      "marginBottom": 3,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Name"
    style={
      {
        "borderColor": "grey",
        "borderRadius": 30,
        "borderWidth": 1,
        "margin": 5,
        "minWidth": "33%",
        "paddingHorizontal": 5,
        "paddingVertical": 0,
      }
    }
  />
  <TextInput
    keyboardType="number-pad"
    onChangeText={[Function]}
    placeholder="Quantity"
    style={
      {
        "borderColor": "grey",
        "borderRadius": 30,
        "borderWidth": 1,
        "margin": 5,
        "minWidth": "33%",
        "paddingHorizontal": 5,
        "paddingVertical": 0,
      }
    }
    value="0"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "justifyContent": "center",
        "opacity": 1,
        "paddingLeft": 10,
        "textAlign": "right",
      }
    }
  >
    <Text>
      Save
    </Text>
  </View>
</View>
`;
